### PR TITLE
[CWS] improve exec exec rebase algorithm CPU usage

### PR DESCRIPTION
**Pull request by paulcacheux**
### What does this PR do?

This PR improves the CPU usage of the exec exec rebase algorithm by:
- improving the allocation patterns
- creating a simple, alloc-free, fast path
- implementing cycle detection (to protect ourselves against a potential bug)
- moving from a queue to a stack when visiting children

### Reviewer's Checklist

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [ ] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the `k8s/<min-version>` label, indicating the lowest Kubernetes version compatible with this feature. 
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

### CODE REVIEW

File: pkg/security/security_profile/activity_tree/activity_tree.go [hunk 1]
```diff
@@ -598,27 +599,29 @@ func (at *ActivityTree) findProcessCacheEntryInTree(tree []*ProcessNode, entry *
...
+
+	visited := make([]*ProcessNode, 0, 64)
 
 	for len(execChildren) > 0 {
-		cursor := execChildren[0]
-		execChildren = execChildren[1:]
+		cursor := execChildren[len(execChildren)-1]
+		execChildren = execChildren[:len(execChildren)-1]
+
+		visited = append(visited, cursor)
```
Comment (paulcacheux) [inline pkg/security/security_profile/activity_tree/activity_tree.go:611]: cycle detection

File: pkg/security/security_profile/activity_tree/activity_tree.go [hunk 2]
```diff
@@ -598,27 +599,29 @@ func (at *ActivityTree) findProcessCacheEntryInTree(tree []*ProcessNode, entry *
 // findProcessCacheEntryInChildExecedNodes look for entry in the execed nodes of child
 func (at *ActivityTree) findProcessCacheEntryInChildExecedNodes(child *ProcessNode, entry *model.ProcessCacheEntry) *ProcessNode {
 	// children is used to iterate over the tree below child
-	execChildren := []*ProcessNode{child}
+	execChildren := make([]*ProcessNode, 1, 64)
```
Comment (paulcacheux) [inline pkg/security/security_profile/activity_tree/activity_tree.go:602]: pre-alloc of slices

File: pkg/security/security_profile/activity_tree/activity_tree.go [hunk 3]
```diff
@@ -598,27 +599,29 @@ func (at *ActivityTree) findProcessCacheEntryInTree(tree []*ProcessNode, entry *
...
 				// there should always be only one
-				execChildren = append(execChildren, node)
-			}
-		}
 
-		if len(execChildren) == 0 {
-			break
-		}
+				// does this execed child match the entry ?
+				if node.Matches(&entry.Process, at.differentiateArgs) {
```
Comment (paulcacheux) [inline pkg/security/security_profile/activity_tree/activity_tree.go:619]: match check in line with iteration (skip iteration if not needed, no alloc if a match is found)

**Review by spikat — APPROVED** [review]

**Comment by modernplumbing** [comment]
did you do some benchmarking on the before and after? would be cool to see

**Review by lebauce — APPROVED** [review]

**Pull request merged**